kyaml/yaml/walk/map.go

```diff
@@ -168,6 +176,9 @@ func (l Walker) fieldValue(fieldName string) ([]*yaml.RNode, *openapi.ResourceSc
 		if sch == nil && !s.IsMissingOrNull() {
 			sch = s
 		}
+		if keyStyle == 0 && field != nil && field.Key != nil && field.Key.YNode() != nil {
+			keyStyle = field.Key.YNode().Style
```

Comment:
I see that this code is inside a loop over `l.Sources`. I'm not very familiar with this part of the code (nor are any of the other active maintainers), but preliminary debugging seems to indicate that it will contain all three relevant KV pairs in a three-way merge (if all three documents have the field). All three corresponding values will be returned, and the final value will be selected by the Walker.Walk call at L66-73. However, this PR is currently selecting the first non-zero style from among the three candidate keys, which may not correspond to the value that is retained in the final document. I feel like that is unexpected.

Having the SMP transformer tests you've already written is great, but I would like to see lower-level test cases for the Walker as well. I believe the merge3 and/or merge2 test suites are the best place for this, but if you find a better one, please point it out.

Here's a merge3 case I used to discover what I explained above:

```go
//
	// Test Case
	//
	{
		description: `The correct key style is used in the result`,
		origin: `
apiVersion: v1
kind: ConfigMap
metadata:
  name: foo
data:
  1: a
`,
		update: `
apiVersion: v1
kind: ConfigMap
metadata:
  name: foo
data:
  "1": b
`,
		local: `
apiVersion: v1
kind: ConfigMap
metadata:
  name: foo
data:
  '1': c
`,
		expected: `
apiVersion: v1
kind: ConfigMap
metadata:
  name: foo
data:
  '1': b` // Current result in this PR: takes the key style from 'local' and the value from 'update' -- unexpected?
},
```

Reply:
Wholly agreed. Thanks for the catch.

Reply:
Similar to the other comment, the behavior described here is true in master - it's not the result of this PR. I'm going to adjust the logic I'm using and will add test cases into merge2 and merge3, but I don't intend to change the current quote style handling for patches against existing keys and values.